src/catalog/src/stash.rs

```diff
@@ -50,148 +52,295 @@ use crate::transaction::{
-            if !stash.is_readonly() {
-                match stash.upgrade().await {
-                    Ok(()) => {}
-                    Err(e) => {
-                        return Err((stash, e.into()));
-                    }
-                }
-            }
+    async fn is_initialized(&mut self) -> Result<bool, Error> {
+        let stash = match &mut self.stash {
+            None => match self.open_stash_read_only().await {
+                Ok(stash) => stash,
+                Err(e) if e.can_recover_with_write_mode() => return Ok(false),
+                Err(e) => return Err(e.into()),
+            },
+            Some(stash) => stash,
+        };
+        stash.is_initialized().await.err_into()
+    }
 
-            let mut conn = Connection { stash };
+    async fn get_deployment_generation(&mut self) -> Result<Option<u64>, Error> {
+        let stash = match &mut self.stash {
+            None => match self.open_stash_read_only().await {
+                Ok(stash) => stash,
+                Err(e) if e.can_recover_with_write_mode() => return Ok(None),
+                Err(e) => return Err(e.into()),
+            },
+            Some(stash) => stash,
+        };
 
-            if !conn.stash.is_readonly() {
-                if let Some(deploy_generation) = deploy_generation {
-                    match conn.set_deploy_generation(deploy_generation).await {
-                        Ok(()) => {}
-                        Err(e) => {
-                            return Err((conn.stash, e));
-                        }
-                    }
+        let deployment_generation = CONFIG_COLLECTION
+            .peek_key_one(
+                stash,
+                proto::ConfigKey {
+                    key: DEPLOY_GENERATION.into(),
+                },
+            )
+            .await?
+            .map(|v| v.value);
+
+        Ok(deployment_generation)
+    }
+}
+
+/// Opens a new [`Connection`] to the stash, retrying any retryable errors. Optionally
+/// initialize the stash if it has not been initialized and perform any migrations needed.
+///
+/// # Panics
+/// If the inner stash has not been opened.
+async fn retry_open(
+    mut stash: Stash,
+    now: NowFn,
+    bootstrap_args: &BootstrapArgs,
+    deploy_generation: Option<u64>,
+) -> Result<Connection, Error> {
+    let retry = Retry::default()
+        .clamp_backoff(Duration::from_secs(1))
+        .max_duration(Duration::from_secs(30))
+        .into_retry_stream();
+    let mut retry = pin::pin!(retry);
+
+    loop {
+        match open_inner(stash, now.clone(), bootstrap_args, deploy_generation).await {
+            Ok(conn) => {
+                return Ok(conn);
+            }
+            Err((given_stash, err)) => {
+                stash = given_stash;
+                let should_retry = matches!(&err, Error::Stash(se) if se.should_retry());
+                if !should_retry || retry.next().await.is_none() {
+                    return Err(err);
```

Comment:
Similar argument, might be interesting to have a test that exercises this error case to see that the error is correctly propagated all the way.

Reply:
This is hard to exercise because we have to mock a CRDB issue which is not easy.